### PR TITLE
Annotation & Alerts: Makes various annotation and alert requests cancelable

### LINKED ISSUE #21943: Builtin annotation queries not cancelled when dashboard refresh
**What happened**:
Refreshed dashboard where builtin annotations is enabled, but the annotation query was not cancelled when refreshing and query not yet finished.

**What you expected to happen**:
Annotation query to be cancelled just like all other datasource queries.

**How to reproduce it (as minimally and precisely as possible)**:
1. Change the code for annotation api query to make it take 5 seconds:
```diff
diff --git a/pkg/api/annotations.go b/pkg/api/annotations.go
index 4720b38..998b89e 100644
--- a/pkg/api/annotations.go
+++ b/pkg/api/annotations.go
@@ -2,6 +2,7 @@ package api
 
 import (
        "strings"
+       "time"
 
        "github.com/grafana/grafana/pkg/api/dtos"
        m "github.com/grafana/grafana/pkg/models"
@@ -11,7 +12,7 @@ import (
 )
 
 func GetAnnotations(c *m.ReqContext) Response {
-
+       time.Sleep(time.Second * 5)
        query := &annotations.ItemQuery{
                From:        c.QueryInt64("from"),
                To:          c.QueryInt64("to"),
```
2. Use a dashboard that have builtin annotations enabled.
3. Refresh the dashboard faster than 5 seconds and you'll notice that annotation queries are not cancelled.

**Anything else we need to know?**:
Was found when investigated #21915 

In addition, annotation_srv also fetches `/api/alerts/states-for-dashboard` and would be good to put a cancel on that one as well.
In addition, alert list panel also fetches annotations and `/api/alerts` and would be good to put a cancel on those as well.

**Environment**:
- Grafana version: v6.6.0
- Data source type & version: Grafana datasource/builtin annotation

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
Adds utility function called `getRequestWithCancel` that is used by various requests to `api/annotations` & `api/alerts`. The important part is the various `requestId`used so would love some feedback on those and the name `getRequestWithCancel` isn't the best so any suggestions would be appreciated.

**Which issue(s) this PR fixes**:
Fixes #21943

**Special notes for your reviewer**:

### CODE REVIEW

**Review by torkelo — APPROVED** [review]
Looks great, found one problem the requestId for /api/alerts/states-for-dashboard call

**Comment by hugohaggmark** [comment]
About to redo this with the new cancellation logic in backendSrv, so no hurry to review this yet @marefr

**Comment by hugohaggmark** [comment]
Ready for review agian @marefr @torkelo

**Review by marefr — COMMENTED** [review]
I don't understand how this can work without changes to backend_srv. What am I missing?

**Comment by hugohaggmark** [comment]
> I don't understand how this can work without changes to backend_srv. What am I missing?

I think you're looking for this #22066 @marefr? Wanted to keep those changes in a separate PR.

File: public/app/features/annotations/annotations_srv.ts [hunk 1]
```diff
@@ -87,9 +84,13 @@ export class AnnotationsSrv {
...
 
-    this.alertStatesPromise = getBackendSrv().get('/api/alerts/states-for-dashboard', {
-      dashboardId: options.dashboard.id,
-    });
+    this.alertStatesPromise = getBackendSrv().get(
+      '/api/alerts/states-for-dashboard',
+      {
+        dashboardId: options.dashboard.id,
+      },
+      `annotation-srv-get-alert-states-${options.dashboard.id}`
```
Comment (marefr) [inline public/app/features/annotations/annotations_srv.ts]: Nit
```suggestion
      `get-alert-states-${options.dashboard.id}`
```

**Review by marefr — APPROVED** [review]
Works as expected :+1: Added a nit-suggestion. LGTM

**Pull request merged**